File: src/org/stringtemplate/v4/gui/STViz.java
/*
 * [The "BSD license"]
 *  Copyright (c) 2011 Terence Parr
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *  1. Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *  2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *  3. The name of the author may not be used to endorse or promote products
 *     derived from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 *  IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 *  OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 *  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 *  NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 *  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 *  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 *  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package org.stringtemplate.v4.gui;

import org.antlr.runtime.CommonToken;
import org.antlr.runtime.tree.CommonTree;
import org.antlr.runtime.tree.CommonTreeAdaptor;
import org.stringtemplate.v4.InstanceScope;
import org.stringtemplate.v4.Interpreter;
import org.stringtemplate.v4.ST;
import org.stringtemplate.v4.STGroup;
import org.stringtemplate.v4.STGroupFile;
import org.stringtemplate.v4.STGroupString;
import org.stringtemplate.v4.debug.EvalExprEvent;
import org.stringtemplate.v4.debug.EvalTemplateEvent;
import org.stringtemplate.v4.debug.InterpEvent;
import org.stringtemplate.v4.misc.ErrorManager;
import org.stringtemplate.v4.misc.Interval;
import org.stringtemplate.v4.misc.Misc;
import org.stringtemplate.v4.misc.STMessage;
import org.stringtemplate.v4.misc.STRuntimeMessage;

import javax.swing.*;
import javax.swing.border.Border;
import javax.swing.event.CaretEvent;
import javax.swing.event.CaretListener;
import javax.swing.event.ListSelectionEvent;
import javax.swing.event.ListSelectionListener;
import javax.swing.event.TreeSelectionEvent;
import javax.swing.event.TreeSelectionListener;
import javax.swing.text.BadLocationException;
import javax.swing.text.DefaultHighlighter;
import javax.swing.text.Highlighter;
import javax.swing.text.JTextComponent;
import javax.swing.tree.TreePath;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

public class STViz {
	protected static final String WINDOWS_LINE_ENDINGS = "WINDOWS_LINE_ENDINGS";

	//public ST currentST; // current ST selected in template tree
	public EvalTemplateEvent root;
	public InterpEvent currentEvent;
	public InstanceScope currentScope;
	public List<InterpEvent> allEvents;
	public JTreeSTModel tmodel;
	public ErrorManager errMgr;
	public Interpreter interp;
	public String output;
	public List<String> trace;
	public List<STMessage> errors;

	public STViewFrame viewFrame;

	private final AtomicInteger updateDepth = new AtomicInteger();

    public STViz(ErrorManager errMgr,
				 EvalTemplateEvent root,
				 String output,
				 Interpreter interp,
                 List<String> trace,
                 List<STMessage> errors)
    {
		this.errMgr = errMgr;
		this.currentEvent = root;
		this.currentScope = root.scope;
		this.output = output;
		this.interp = interp;
        this.allEvents = interp.getEvents();
		this.trace = trace;
        this.errors = errors;
	}

	public void open() {
        viewFrame = new STViewFrame();
        updateStack(currentScope, viewFrame);
        updateAttributes(currentScope, viewFrame);

		List<InterpEvent> events = currentScope.events;
		tmodel = new JTreeSTModel(interp, (EvalTemplateEvent)events.get(events.size()-1));
        viewFrame.tree.setModel(tmodel);
        viewFrame.tree.addTreeSelectionListener(
            new TreeSelectionListener() {
				@Override
                public void valueChanged(TreeSelectionEvent treeSelectionEvent) {
					int depth = updateDepth.incrementAndGet();
					try {
						if (depth != 1) {
							return;
						}

						currentEvent = ((JTreeSTModel.Wrapper)viewFrame.tree.getLastSelectedPathComponent()).event;
						currentScope = currentEvent.scope;
						updateCurrentST(viewFrame);
					}
					finally {
						updateDepth.decrementAndGet();
					}
                }
            }
        );

		JTreeASTModel astModel = new JTreeASTModel(new CommonTreeAdaptor(), currentScope.st.impl.ast);
		viewFrame.ast.setModel(astModel);
		viewFrame.ast.addTreeSelectionListener(
			new TreeSelectionListener() {
				@Override
				public void valueChanged(TreeSelectionEvent treeSelectionEvent) {
					int depth = updateDepth.incrementAndGet();
					try {
						if (depth != 1) {
							return;
						}

						TreePath path = treeSelectionEvent.getNewLeadSelectionPath();
						if ( path==null ) return;
						CommonTree node = (CommonTree)treeSelectionEvent.getNewLeadSelectionPath().getLastPathComponent();
						//System.out.println("select AST: "+node);
						CommonToken a = (CommonToken)currentScope.st.impl.tokens.get(node.getTokenStartIndex());
						CommonToken b = (CommonToken)currentScope.st.impl.tokens.get(node.getTokenStopIndex());
						highlight(viewFrame.template, a.getStartIndex(), b.getStopIndex());
					}
					finally {
						updateDepth.decrementAndGet();
					}
				}
			}
		);

		// Track selection of attr but do nothing for now
//        viewFrame.attributes.addListSelectionListener(
//            new ListSelectionListener() {
//                public void valueChanged(ListSelectionEvent e) {
//                    int minIndex = viewFrame.attributes.getMinSelectionIndex();
//                    int maxIndex = viewFrame.attributes.getMaxSelectionIndex();
//                    for (int i = minIndex; i <= maxIndex; i++) {
//                        if (viewFrame.attributes.isSelectedIndex(i)) {
//                            //System.out.println("index="+i);
//                        }
//                    }
//                }
//            }
//        );

        CaretListener caretListenerLabel = new CaretListener() {
			@Override
            public void caretUpdate(CaretEvent e) {
				int depth = updateDepth.incrementAndGet();
				try {
					if (depth != 1) {
						return;
					}

					int dot = toEventPosition((JTextComponent)e.getSource(), e.getDot());
					currentEvent = findEventAtOutputLocation(allEvents, dot);
					if ( currentEvent==null ) currentScope = tmodel.root.event.scope;
					else currentScope = currentEvent.scope;

					// update tree view of template hierarchy
					// compute path from root to currentST, create TreePath for tree widget
					List<EvalTemplateEvent> stack = Interpreter.getEvalTemplateEventStack(currentScope, true);
					//System.out.println("\nselect path="+stack);
					Object[] path = new Object[stack.size()];
					int j = 0;
					for (EvalTemplateEvent s : stack) {
						path[j++] = new JTreeSTModel.Wrapper(s);
					}
					TreePath p = new TreePath(path);
					viewFrame.tree.setSelectionPath(p);
					viewFrame.tree.scrollPathToVisible(p);
					updateCurrentST(viewFrame);
				}
				finally {
					updateDepth.decrementAndGet();
				}
			}
		};

		viewFrame.output.addCaretListener(caretListenerLabel);

        // ADD ERRORS
        if ( errors==null || errors.size()==0 ) {
            viewFrame.errorScrollPane.setVisible(false); // don't show unless errors
        }
        else {
            final DefaultListModel errorListModel = new DefaultListModel();
            for (STMessage msg : errors) {
                errorListModel.addElement(msg);
            }
            viewFrame.errorList.setModel(errorListModel);
        }

        viewFrame.errorList.addListSelectionListener(
            new ListSelectionListener() {
				@Override
                public void valueChanged(ListSelectionEvent e) {
					int depth = updateDepth.incrementAndGet();
					try {
						if (depth != 1) {
							return;
						}

						int minIndex = viewFrame.errorList.getMinSelectionIndex();
						int maxIndex = viewFrame.errorList.getMaxSelectionIndex();
						int i = minIndex;
						while ( i <= maxIndex ) {
							if (viewFrame.errorList.isSelectedIndex(i)) break;
							i++;
						}
						ListModel model = viewFrame.errorList.getModel();
						STMessage msg = (STMessage)model.getElementAt(i);
						if ( msg instanceof STRuntimeMessage ) {
							STRuntimeMessage rmsg = (STRuntimeMessage)msg;
							Interval I = rmsg.self.impl.sourceMap[rmsg.ip];
							currentEvent = null;
							currentScope = ((STRuntimeMessage) msg).scope;
							updateCurrentST(viewFrame);
							if ( I!=null ) { // highlight template
								highlight(viewFrame.template, I.a, I.b);
							}
						}
					}
					finally {
						updateDepth.decrementAndGet();
					}
                }
            }
        );

		Border empty = BorderFactory.createEmptyBorder();
		viewFrame.treeContentSplitPane.setBorder(empty);
		viewFrame.outputTemplateSplitPane.setBorder(empty);
		viewFrame.templateBytecodeTraceTabPanel.setBorder(empty);
		viewFrame.treeAttributesSplitPane.setBorder(empty);


		viewFrame.treeContentSplitPane.setOneTouchExpandable(true);
		viewFrame.outputTemplateSplitPane.setOneTouchExpandable(true);
		viewFrame.treeContentSplitPane.setDividerSize(10);
		viewFrame.outputTemplateSplitPane.setDividerSize(8);
		viewFrame.treeContentSplitPane.setContinuousLayout(true);
		viewFrame.treeAttributesSplitPane.setContinuousLayout(true);
		viewFrame.outputTemplateSplitPane.setContinuousLayout(true);

		viewFrame.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
		viewFrame.pack();
		viewFrame.setSize(900, 700);

        setText(viewFrame.output, output);
        setText(viewFrame.template, currentScope.st.impl.template);
        setText(viewFrame.bytecode, currentScope.st.impl.disasm());
        setText(viewFrame.trace, Misc.join(trace.iterator(), "\n"));

        viewFrame.setVisible(true);
    }

	public void waitForClose() throws InterruptedException {
		final Object lock = new Object();

		Thread t = new Thread() {
			@Override
			public void run() {
				synchronized (lock) {
					while (viewFrame.isVisible()) {
						try {
							lock.wait();
						} catch (InterruptedException e) {
						}
					}
				}
			}
		};

		t.start();

		viewFrame.addWindowListener(new WindowAdapter() {
			@Override
			public void windowClosing(WindowEvent arg0) {
				synchronized (lock) {
					viewFrame.setVisible(false);
					lock.notify();
				}
			}
		});

		t.join();
	}

	private void updateCurrentST(STViewFrame m) {
//		System.out.println("updateCurrentST(): currentScope.st="+currentScope.st);
		// update all views according to currentScope.st
		updateStack(currentScope, m); 					   // STACK
		updateAttributes(currentScope, m); 			 	   // ATTRIBUTES
        setText(m.bytecode, currentScope.st.impl.disasm()); // BYTECODE DIS.
		setText(m.template, currentScope.st.impl.template); // TEMPLATE SRC
		JTreeASTModel astModel = new JTreeASTModel(new CommonTreeAdaptor(), currentScope.st.impl.ast);
		viewFrame.ast.setModel(astModel);

		// highlight output text and, if {...} subtemplate, region in ST src
		// get last event for currentScope.st; it's the event that captures ST eval
		if (currentEvent instanceof EvalExprEvent) {
			EvalExprEvent exprEvent = (EvalExprEvent)currentEvent;
			highlight(m.output, exprEvent.outputStartChar, exprEvent.outputStopChar);
			highlight(m.template, exprEvent.exprStartChar, exprEvent.exprStopChar);
		}
		else {
			EvalTemplateEvent templateEvent;
			if (currentEvent instanceof EvalTemplateEvent) {
				templateEvent = (EvalTemplateEvent)currentEvent;
			}
			else {
				List<InterpEvent> events = currentScope.events;
				templateEvent = (EvalTemplateEvent)events.get(events.size() - 1);
			}

			if (templateEvent != null) {
				highlight(m.output, templateEvent.outputStartChar, templateEvent.outputStopChar);
			}

			if ( currentScope.st.isAnonSubtemplate() ) {
				Interval r = currentScope.st.impl.getTemplateRange();
				//System.out.println("currentScope.st src range="+r);
				//m.template.moveCaretPosition(r.a);
				highlight(m.template, r.a, r.b);
			}
		}
	}

	protected void setText(JEditorPane component, String text) {
		List<Integer> windowsLineEndingsList = new ArrayList<Integer>();
		for (int i = 0; i < text.length(); i += 2) {
			i = text.indexOf("\r\n", i);
			if (i < 0) {
				break;
			}

			windowsLineEndingsList.add(i);
		}

		int[] windowsLineEndings = new int[windowsLineEndingsList.size()];
		for (int i = 0; i < windowsLineEndingsList.size(); i++) {
			windowsLineEndings[i] = windowsLineEndingsList.get(i);
		}

		component.setText(text);
		component.getDocument().putProperty(WINDOWS_LINE_ENDINGS, windowsLineEndings);
	}

	protected int toComponentPosition(JTextComponent component, int position) {
		int[] windowsLineEndings = (int[])component.getDocument().getProperty(WINDOWS_LINE_ENDINGS);
		if (windowsLineEndings == null || windowsLineEndings.length == 0) {
			return position;
		}

		int index = Arrays.binarySearch(windowsLineEndings, position);
		if (index >= 0) {
			return position - index;
		}

		return position - (-index - 1);
	}

	protected int toEventPosition(JTextComponent component, int position) {
		int result = position;
		while (toComponentPosition(component, result) < position) {
			result++;
		}

		return result;
	}

	protected final void highlight(JTextComponent comp, int i, int j) {
		highlight(comp, i, j, true);
	}

	protected void highlight(JTextComponent comp, int i, int j, boolean scroll) {
		Highlighter highlighter = comp.getHighlighter();
		highlighter.removeAllHighlights();

		try {
			i = toComponentPosition(comp, i);
			j = toComponentPosition(comp, j);
			highlighter.addHighlight(i, j+1, DefaultHighlighter.DefaultPainter);
			if (scroll) {
				if (comp.getCaretPosition() < i || comp.getCaretPosition() > j) {
					comp.moveCaretPosition(i);
					comp.scrollRectToVisible(comp.modelToView(i));
				}
			}
		}
		catch (BadLocationException ble) {
			errMgr.internalError(tmodel.root.event.scope.st, "bad highlight location", ble);
		}
	}

	protected void updateAttributes(final InstanceScope scope, final STViewFrame m) {
		//System.out.println("updateAttributes: "+Interpreter.getEnclosingInstanceStackString(scope) );
		m.attributes.setModel( new JTreeScopeStackModel(scope) );
		m.attributes.setRootVisible(false);
		m.attributes.setShowsRootHandles(true);
		//System.out.println("add events="+ st.addAttrEvents);
//		ST st = scope.st;
//		final DefaultListModel attrModel = new DefaultListModel();
//		final Map<String,Object> attrs = st.getAttributes();
//		if ( attrs!=null ) {
//			for (String a : attrs.keySet()) {
//				if ( st.debugState!=null && st.debugState.addAttrEvents!=null ) {
//					List<AddAttributeEvent> events = st.debugState.addAttrEvents.get(a);
//					StringBuilder locations = new StringBuilder();
//					int i = 0;
//					if ( events!=null ) {
//						for (AddAttributeEvent ae : events) {
//							if ( i>0 ) locations.append(", ");
//							locations.append(ae.getFileName()+":"+ae.getLine());
//							i++;
//						}
//					}
//					if ( locations.length()>0 ) {
//						attrModel.addElement(a+" = "+attrs.get(a)+" @ "+locations.toString());
//					}
//					else {
//						attrModel.addElement(a+" = "+attrs.get(a));
//					}
//				}
//				else {
//					attrModel.addElement(a+" = "+attrs.get(a));
//				}
//			}
//		}
//		m.attributes.setModel(attrModel);
	}

	protected void updateStack(InstanceScope scope, STViewFrame m) {
		List<ST> stack = Interpreter.getEnclosingInstanceStack(scope, true);
		m.setTitle("STViz - ["+ Misc.join(stack.iterator()," ")+"]");
//        // also do source stack
//        StackTraceElement[] trace = st.newSTEvent.stack.getStackTrace();
//        StringWriter sw = new StringWriter();
//        for (StackTraceElement e : trace) {
//            sw.write(e.toString()+"\n");
//        }
    }

    public InterpEvent findEventAtOutputLocation(List<InterpEvent> events,
                                                 int charIndex)
    {
        for (InterpEvent e : events) {
			if (e.scope.earlyEval) {
				continue;
			}

            if ( charIndex>=e.outputStartChar && charIndex<=e.outputStopChar) return e;
        }
        return null;
	}

    public static void main(String[] args) throws IOException { // test rig
		if ( args.length>0 && args[0].equals("1") ) test1();
		else if ( args.length>0 && args[0].equals("2") ) test2();
		else if ( args.length>0 && args[0].equals("3") ) test3();
		else if ( args.length>0 && args[0].equals("4") ) test4();
	}

	public static void test1() throws IOException { // test rig
        String templates =
			"method(type,name,locals,args,stats) ::= <<\n" +
			"public <type> <name>(<args:{a| int <a>}; separator=\", \">) {\n" +
			"    <if(locals)>int locals[<locals>];<endif>\n"+
			"    <stats;separator=\"\\n\">\n" +
			"}\n" +
			">>\n"+
			"assign(a,b) ::= \"<a> = <b>;\"\n"+
			"return(x) ::= <<return <x>;>>\n" +
			"paren(x) ::= \"(<x>)\"\n";

        String tmpdir = System.getProperty("java.io.tmpdir");
        writeFile(tmpdir, "t.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/"+"t.stg");
        ST st = group.getInstanceOf("method");
        st.impl.dump();
        st.add("type", "float");
        st.add("name", "foo");
        st.add("locals", 3);
        st.add("args", new String[] {"x", "y", "z"});
        ST s1 = group.getInstanceOf("assign");
        ST paren = group.getInstanceOf("paren");
        paren.add("x", "x");
        s1.add("a", paren);
        s1.add("b", "y");
        ST s2 = group.getInstanceOf("assign");
        s2.add("a", "y");
        s2.add("b", "z");
        ST s3 = group.getInstanceOf("return");
        s3.add("x", "3.14159");
        st.add("stats", s1);
        st.add("stats", s2);
        st.add("stats", s3);

		STViz viz = st.inspect();
		System.out.println(st.render()); // should not mess up ST event lists
    }

	public static void test2() throws IOException { // test rig
        String templates =
			"t1(q1=\"Some\\nText\") ::= <<\n" +
			"<q1>\n" +
			">>\n" +
			"\n" +
			"t2(p1) ::= <<\n" +
			"<p1>\n" +
			">>\n" +
			"\n" +
			"main() ::= <<\n" +
			"START-<t1()>-END\n" +
			"\n" +
			"START-<t2(p1=\"Some\\nText\")>-END\n" +
			">>\n";

        String tmpdir = System.getProperty("java.io.tmpdir");
        writeFile(tmpdir, "t.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/"+"t.stg");
        ST st = group.getInstanceOf("main");
        STViz viz = st.inspect();
    }

	public static void test3() throws IOException {
        String templates =
			"main() ::= <<\n" +
			"Foo: <{bar};format=\"lower\">\n" +
			">>\n";

        String tmpdir = System.getProperty("java.io.tmpdir");
        writeFile(tmpdir, "t.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/"+"t.stg");
        ST st = group.getInstanceOf("main");
        st.inspect();
    }

	public static void test4() throws IOException {
        String templates =
			"main(t) ::= <<\n" +
			"hi: <t>\n" +
			">>\n" +
			"foo(x,y={hi}) ::= \"<bar(x,y)>\"\n" +
			"bar(x,y) ::= << <y> >>\n" +
			"ignore(m) ::= \"<m>\"\n";

        STGroup group = new STGroupString(templates);
        ST st = group.getInstanceOf("main");
		ST foo = group.getInstanceOf("foo");
		st.add("t", foo);
		ST ignore = group.getInstanceOf("ignore");
		ignore.add("m", foo); // embed foo twice!
        st.inspect();
		st.render();
    }

    public static void writeFile(String dir, String fileName, String content) {
        try {
            File f = new File(dir, fileName);
            if ( !f.getParentFile().exists() ) f.getParentFile().mkdirs();
            FileWriter w = new FileWriter(f);
            BufferedWriter bw = new BufferedWriter(w);
            bw.write(content);
            bw.close();
            w.close();
        }
        catch (IOException ioe) {
            System.err.println("can't write file");
            ioe.printStackTrace(System.err);
        }
    }
}

File: objc/main/gui/STViz.m
#import "STViz.h"

@implementation STViz_Anon1

- (void) valueChanged:(TreeSelectionEvent *)treeSelectionEvent
{
    currentScope = ((Wrapper *)[viewFrame.tree lastSelectedPathComponent]).event.scope;
    [self updateCurrentST:viewFrame];
}

@end

@implementation STViz_Anon2

- (void) valueChanged:(TreeSelectionEvent *)treeSelectionEvent
{
    TreePath * path = [treeSelectionEvent newLeadSelectionPath];
    if (path == nil)
        return;
    CommonTree *node = (CommonTree *)[[treeSelectionEvent newLeadSelectionPath] lastPathComponent];
    CommonToken *a = (CommonToken *)[currentScope.st.impl.tokens get:[node tokenStartIndex]];
    CommonToken *b = (CommonToken *)[currentScope.st.impl.tokens get:[node tokenStopIndex]];
    [self highlight:viewFrame.template i:[a startIndex] j:[b stopIndex]];
}

@end

@implementation STViz_Anon3

- (void) caretUpdate:(CaretEvent *)e
{
    NSInteger dot = [e dot];
    InterpEvent *de = [self findEventAtOutputLocation:allEvents param1:dot];
    if (de == nil)
        currentScope = tmodel.root.event.scope;
    else
        currentScope = de.scope;
    NSMutableArray *stack = [Interpreter getEvalTemplateEventStack:currentScope param1:YES];
    NSArray *path = [NSArray array];
    int j = 0;
    
    for (EvalTemplateEvent *s in stack)
        path[j++] = [[[Wrapper alloc] init:s] autorelease];
    
    TreePath *p = [[[TreePath alloc] init:path] autorelease];
    [viewFrame.tree setSelectionPath:p];
    [viewFrame.tree scrollPathToVisible:p];
    [self updateCurrentST:viewFrame];
}

@end

@implementation STViz_Anon4

- (void) valueChanged:(ListSelectionEvent *)e
{
    NSInteger minIndex = [viewFrame.errorList minSelectionIndex];
    NSInteger maxIndex = [viewFrame.errorList maxSelectionIndex];
    NSInteger i = minIndex;
    
    while (i <= maxIndex) {
        if ([viewFrame.errorList isSelectedIndex:i])
            break;
        i++;
    }
    
    ListModel *model = [viewFrame.errorList model];
    STMessage *msg = (STMessage *)[model getElementAt:i];
    if ([msg conformsToProtocol:@protocol(STRuntimeMessage)]) {
        STRuntimeMessage *rmsg = (STRuntimeMessage *)msg;
        Interval *I = rmsg.self.impl.sourceMap[rmsg.ip];
        currentScope = ((STRuntimeMessage *)msg).scope;
        [self updateCurrentST:viewFrame];
        if (I != nil) {
            [self highlight:viewFrame.template i:I.a j:I.b];
        }
    }
}

@end

@implementation STViz

+ (STViz *) newSTViz:(ErrorManager *)anErrMgr root:(EvalTemplateEvent *)aRoot output:(NSString *)anOutput interp:(Interpreter *)anInterp trace:(AMutableArray *)aTrace errors:(AMutableArray *)theErrors
{
    return [[STViz alloc] init:anErrMgr root:aRoot output:anOutput interp:anInterp trace:aTrace errors:theErrors];
}

- (id) init:(ErrorManager *)anErrMgr root:(EvalTemplateEvent *)aRoot output:(NSString *)anOutput interp:(Interpreter *)anInterp trace:(AMutableArray *)aTrace errors:(AMutableArray *)theErrors
{
    if (self = [super init]) {
        errMgr = errMgr;
        currentScope = root.scope;
        output = output;
        interp = interp;
        allEvents = [interp events];
        trace = trace;
        errors = errors;
    }
    return self;
}

- (void) open
{
    viewFrame = [[[STViewFrame alloc] init] autorelease];
    [self updateStack:currentScope m:viewFrame];
    [self updateAttributes:currentScope m:viewFrame];
    NSMutableArray * events = currentScope.events;
    tmodel = [[[JTreeSTModel alloc] init:interp param1:(EvalTemplateEvent *)[events objectAtIndex:[events count] - 1]] autorelease];
    [viewFrame.tree setModel:tmodel];
    [viewFrame.tree addTreeSelectionListener:[[[STViz_Anon1 alloc] init] autorelease]];
    JTreeASTModel * astModel = [[[JTreeASTModel alloc] init:[[[CommonTreeAdaptor alloc] init] autorelease] param1:currentScope.st.impl.ast] autorelease];
    [viewFrame.ast setModel:astModel];
    [viewFrame.ast addTreeSelectionListener:[[[STViz_Anon2 alloc] init] autorelease]];
    [viewFrame.output setText:output];
    [viewFrame.template setText:currentScope.st.impl.template];
    [viewFrame.bytecode setText:[currentScope.st.impl disasm]];
    [viewFrame.trace setText:[Misc join:[trace objectEnumerator] param1:@"\n"]];
    CaretListener * caretListenerLabel = [[[STViz_Anon3 alloc] init] autorelease];
    [viewFrame.output addCaretListener:caretListenerLabel];
    if (errors == nil || [errors count] == 0) {
        [viewFrame.errorScrollPane setVisible:NO];
    }
    else {
        DefaultListModel *errorListModel = [[[DefaultListModel alloc] init] autorelease];
        
        for (STMessage *msg in errors) {
            [errorListModel addElement:msg];
        }
        
        [viewFrame.errorList setModel:errorListModel];
    }
    [viewFrame.errorList addListSelectionListener:[[[STViz_Anon4 alloc] init] autorelease]];
    Border * empty = [BorderFactory createEmptyBorder];
    [viewFrame.treeContentSplitPane setBorder:empty];
    [viewFrame.outputTemplateSplitPane setBorder:empty];
    [viewFrame.templateBytecodeTraceTabPanel setBorder:empty];
    [viewFrame.treeAttributesSplitPane setBorder:empty];
    [viewFrame.treeContentSplitPane setOneTouchExpandable:YES];
    [viewFrame.outputTemplateSplitPane setOneTouchExpandable:YES];
    [viewFrame.treeContentSplitPane setDividerSize:10];
    [viewFrame.outputTemplateSplitPane setDividerSize:8];
    [viewFrame.treeContentSplitPane setContinuousLayout:YES];
    [viewFrame.treeAttributesSplitPane setContinuousLayout:YES];
    [viewFrame.outputTemplateSplitPane setContinuousLayout:YES];
    [viewFrame setDefaultCloseOperation:JFrame.DISPOSE_ON_CLOSE];
    [viewFrame pack];
    [viewFrame setSize:900 param1:700];
    [viewFrame setVisible:YES];
}

- (void) updateCurrentST:(STViewFrame *)m
{
    [self updateStack:currentScope m:m];
    [self updateAttributes:currentScope m:m];
    [m.bytecode moveCaretPosition:0];
    [m.bytecode setText:[currentScope.st.impl disasm]];
    [m.template moveCaretPosition:0];
    [m.template setText:currentScope.st.impl.template];
    JTreeASTModel * astModel = [[[JTreeASTModel alloc] init:[[[CommonTreeAdaptor alloc] init] autorelease] param1:currentScope.st.impl.ast] autorelease];
    [viewFrame.ast setModel:astModel];
    NSMutableArray * events = currentScope.events;
    EvalTemplateEvent * e = (EvalTemplateEvent *)[events objectAtIndex:[events count] - 1];
    [self highlight:m.output i:e.outputStartChar j:e.outputStopChar];
    
    @try {
        [m.output scrollRectToVisible:[m.output modelToView:e.outputStartChar]];
    }
    @catch (BadLocationException * ble) {
        [currentScope.st.groupThatCreatedThisInstance.errMgr internalError:currentScope.st param1:[@"bad location: char index " stringByAppendingString:e.outputStartChar] param2:ble];
    }
    if ([currentScope.st anonSubtemplate]) {
        Interval * r = [currentScope.st.impl templateRange];
        [self highlight:m.template i:r.a j:r.b];
    }
}

- (void) highlight:(JTextComponent *)comp i:(int)i j:(int)j
{
    Highlighter * highlighter = [comp highlighter];
    [highlighter removeAllHighlights];
    
    @try {
        [highlighter addHighlight:i param1:j + 1 param2:DefaultHighlighter.DefaultPainter];
    }
    @catch (BadLocationException * ble) {
        [errMgr internalError:tmodel.root.event.scope.st param1:@"bad highlight location" param2:ble];
    }
}

- (void) updateAttributes:(InstanceScope *)scope m:(STViewFrame *)m {
    [m.attributes setModel:[[[JTreeScopeStackModel alloc] init:scope] autorelease]];
    [m.attributes setRootVisible:NO];
    [m.attributes setShowsRootHandles:YES];
}

- (void) updateStack:(InstanceScope *)scope m:(STViewFrame *)m {
    NSMutableArray * stack = [Interpreter getEnclosingInstanceStack:scope param1:YES];
    [m setTitle:[[@"STViz - [" stringByAppendingString:[Misc join:[stack objectEnumerator] param1:@" "]] stringByAppendingString:@"]"]];
}

- (InterpEvent *) findEventAtOutputLocation:(NSMutableArray *)events charIndex:(int)charIndex {
    
    for (InterpEvent * e in events) {
        if (charIndex >= e.outputStartChar && charIndex <= e.outputStopChar)
            return e;
    }
    
    return nil;
}

+ (void) main:(NSArray *)args {
    if (args.length > 0 && [args[0] isEqualTo:@"1"])
        [self test1];
    else if (args.length > 0 && [args[0] isEqualTo:@"2"])
        [self test2];
    else if (args.length > 0 && [args[0] isEqualTo:@"3"])
        [self test3];
    else if (args.length > 0 && [args[0] isEqualTo:@"4"])
        [self test4];
}

+ (void) test1 {
    NSString * templates = @"method(type,name,locals,args,stats) ::= <<\npublic <type> <name>(<args:{a| int <a>}; separator=\", \">) {\n    <if(locals)>int locals[<locals>];<endif>\n    <stats;separator=\"\\n\">\n}\n>>\nassign(a,b) ::= \"<a> = <b>;\"\nreturn(x) ::= <<return <x>;>>\nparen(x) ::= \"(<x>)\"\n";
    NSString *tmpdir = @"/tmp";
    [self writeFile:tmpdir fileName:@"t.stg" content:templates];
    STGroup * group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/t.stg", tmpdir]] autorelease];
    ST * st = [group getInstanceOf:@"method"];
    [st.impl dump];
    [st add:@"type" value:@"float"];
    [st add:@"name" value:@"foo"];
    [st add:@"locals" value:3];
    [st add:@"args" value:(NSObject *)[NSArray arrayWithObjects:@"x", @"y", @"z", nil]];
    ST * s1 = [group getInstanceOf:@"assign"];
    ST * paren = [group getInstanceOf:@"paren"];
    [paren add:@"x" value:@"x"];
    [s1 add:@"a" value:paren];
    [s1 add:@"b" value:@"y"];
    ST * s2 = [group getInstanceOf:@"assign"];
    [s2 add:@"a" value:@"y"];
    [s2 add:@"b" value:@"z"];
    ST * s3 = [group getInstanceOf:@"return"];
    [s3 add:@"x" value:@"3.14159"];
    [st add:@"stats" value:s1];
    [st add:@"stats" value:s2];
    [st add:@"stats" value:s3];
    STViz * viz = [st inspect];
    [System.out println:[st render]];
}

+ (void) test2 {
    NSString *templates = @"t1(q1=\"Some\\nText\") ::= <<\n<q1>\n>>\n\nt2(p1) ::= <<\n<p1>\n>>\n\nmain() ::= <<\nSTART-<t1()>-END\n\nSTART-<t2(p1=\"Some\\nText\")>-END\n>>\n";
    NSString *tmpdir = @"/tmp";
    [self writeFile:tmpdir fileName:@"t.stg" content:templates];
    STGroup *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/t.stg", tmpdir]] autorelease];
    ST *st = [group getInstanceOf:@"main"];
    STViz * viz = [st inspect];
}

+ (void) test3 {
    NSString *templates = @"main() ::= <<\nFoo: <{bar};format=\"lower\">\n>>\n";
    NSString *tmpdir = [System getProperty:@"java.io.tmpdir"];
    [self writeFile:tmpdir fileName:@"t.stg" content:templates];
    STGroup *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/t.stg", tmpdir]] autorelease];
    ST *st = [group getInstanceOf:@"main"];
    [st inspect];
}

+ (void) test4 {
    NSString *templates = @"main(t) ::= <<\nhi: <t>\n>>\nfoo(x,y={hi}) ::= \"<bar(x,y)>\"\nbar(x,y) ::= << <y> >>\nignore(m) ::= \"<m>\"\n";
    STGroup *group = [[STGroupString newSTGroupString:templates] autorelease];
    ST *st = [group getInstanceOf:@"main"];
    ST *foo = [group getInstanceOf:@"foo"];
    [st add:@"t" value:foo];
    ST *ignore = [group getInstanceOf:@"ignore"];
    [ignore add:@"m" value:foo];
    [st inspect];
    [st render];
}

+ (void) writeFile:(NSString *)dir fileName:(NSString *)fileName content:(NSString *)content {
    
    @try {
        File * f = [[[File alloc] init:dir param1:fileName] autorelease];
        if (![[f parentFile] exists])
            [[f parentFile] mkdirs];
        FileWriter * w = [[[FileWriter alloc] init:f] autorelease];
        BufferedWriter * bw = [[[BufferedWriter alloc] init:w] autorelease];
        [bw write:content];
        [bw close];
        [w close];
    }
    @catch (IOException * ioe) {
        [System.err println:@"can't write file"];
        [ioe printStackTrace:System.err];
    }
}

- (void) dealloc {
    [root release];
    [currentScope release];
    [allEvents release];
    [tmodel release];
    [errMgr release];
    [interp release];
    [output release];
    [trace release];
    [errors release];
    [viewFrame release];
    [super dealloc];
}

@end
